### ISSUE TITLE
References

### ISSUE THREAD

**Issue by simison** [Trustroots/trustroots#718]
Members should be able to leave references for each other.

_This is as a high-level thread to keep track of all the plans around this feature. Other issues and PRs should be linked here for reference and detailed feature discussions should go into those instead._

---

# Implementation

UI for references will live inside profile pages:

<img width="300" alt="image" src="https://user-images.githubusercontent.com/87168/47045264-0fcc8d80-d19b-11e8-96ff-e1635d76d841.png">

Technically we'll call these "user references" or just "references", in contrast to "thread references" which are visible next to message threads for bigger screens:

<img width="250" alt="image" src="https://user-images.githubusercontent.com/87168/47040412-dd1c9800-d18e-11e8-857c-9009610d3863.png">

During the 1st implementation, we'll probably want to merge some parts to continue working on other parts. We can hide the feature using "feature flags", which is just an object in `config/env/*.js` files with `true/false` values.

We can divide the work roughly into four parts and probably into smaller sub-parts. Writing a references API+UI and reading references API+UI. Smaller the PRs the easier reviews, and quicker we can work on this and multiple people can work together at the same time.

There's some prior experimental work done in https://github.com/Trustroots/trustroots/pull/649

Preparation work before 1st iteration: https://github.com/Trustroots/trustroots/pull/719

#### User can leave a reference to anyone:
- They're _not_ required to have message exchange first
- Both need to have "email confirmed"; `user.public` has to be `true`

#### References contain:
- _"How do you know them?"_ -info:
    - _"Met in person"_ (level 1)
    - _"I hosted them"_ (level 2)
    - _"They hosted me"_ (level 3)

    They can choose 1-to-many, but they _have_ to choose at least one or they cannot leave a reference.

    <img width="575" alt="screen shot 2018-10-16 at 23 35 48" src="https://user-images.githubusercontent.com/87168/47046287-7488e780-d19d-11e8-95d7-15d2cb6ab58f.png">

- Recommendation: depending what level they choose, question differs on next level:
    - _"Would you recommend others to meet them?"_ (level 1)
    - _"Would you recommend others to host them?"_ (level 2)
    - _"Would you recommend others to stay with them?"_ (level 3)

    In the DB we can still save this just as one `recommend: true|false`.

    <img width="569" alt="screen shot 2018-10-16 at 23 35 56" src="https://user-images.githubusercontent.com/87168/47046337-8e2a2f00-d19d-11e8-82e6-5877501145cd.png">

#### After writing the reference
- They cannot remove or modify it
- Reference won't become visible at receiving member's profile right away
- Receiving member gets an email + notification about new reference, and that they can reply it within 14 days (we could consider changing this to 3 or 4 weeks)
- If receiver they did not respond with reference in given timeframe, in the response reference they cannot choose "No" or "I don't know" as recommendation options. They can leave only "yes" responses and they'll be shown "Ignore" button for this question in place of "No/I don't know".
- After given timeframe, or after both have left reference, both references become public.

Both reading and writing would ideally be mobile-friendly alrady at this point.

_We'll release 1st iteration already for everyone._

#### Reading references
- The list should be pretty simple at this point, most importantly highlighting the amount of references user has received
- If we can implement it easily, somehow visually show if user has negative references

#### Leaving references
- [ ] Users get more notifications about time they have left to respond to references

- [ ] "No" answers can be reported to moderators
    <img width="869" alt="screen shot 2018-10-16 at 23 36 13" src="https://user-images.githubusercontent.com/87168/47047107-68058e80-d19f-11e8-9643-cd0895cee5a6.png">

- [ ] Users can endorse each other's tribes, choosing up to 3 they think really represents member well. They can endorse only when they choose to recommend member.
    <img width="876" alt="screen shot 2018-10-16 at 23 39 00" src="https://user-images.githubusercontent.com/87168/47047116-6e940600-d19f-11e8-9646-b44d97d80b3f.png">

- [ ] Users can write textual references to each other, but this shouldn't be highlighted as main point of the reference. In fact it should be totally normal, even encouraged, to leave references without _any_ textual reference!
    <img width="869" alt="screen shot 2018-10-17 at 00 14 53" src="https://user-images.githubusercontent.com/87168/47048016-d8adaa80-d1a1-11e8-9992-06f1bbbb8928.png">

#### Reading references
- [ ] Endorsed tribes are listed first at user's profile
- [ ] Number of endorsements is shown at users's profile somehow
- Textual references should _not_ play big role in most of the references. We could highlight them only for negative references?

## Future iterations
- [ ] Users can modify their references until they become public
- [ ] If in future we implement custom "hosting requests", we can somehow make references take those into account but for now more flexibility allows us to learn more.
- [ ] "Add them as a contact" during leaving a reference -flow.

## Likely not implementing, but just listing here for the record
- "I currently live with this person" toggle in first question; receiver needs to confirm it and it'll be added to both profiles as info. They can modify it later on.
- "This person is creepy with women" -checkbox and replies would be shown only to females

# Design questions
- How should reading references look like in first iteration? Next iterations?
- How could we show what types of references user has received?
  - at their profile
  - at the message thread
- How could we show tribe endorsements in profile?
- How should endorsing tribes list look like in 2nd iteration? (1st iteration, albeit a bit uggly, already implemented)
- How should leaving text references look like?
- What should the text in emails and push notifications look like?
- At what points we should send emails and push notifications?
   - When user receives a reference?
   - When user has not replied to reference but there is still one week time to do so, and again two days before limit?
   - After limit is reached?

# Feedback & ideas

Thoughts? Ideas? Experiences?

Please keep discussion constructive and remember not to think only your own use-case but everyone's. References are pretty complex feature serving multiple purposes which all have different needs — they:

- Indicate someone's experience/activity within the community
- A way to say "thank you"
- Show someone if someone is likeable or not; are they generally not liked or are they just different than me?
- Security: are they dangerous? These cases generally shouldn't be covered by references but should be reported to admins instead.

We've put a lot of thinking into this but it's still only theory. We cannot make it perfect from the start and our network functions quite differently from others — we have different constraints to work with (e.g. we don't have "couch requests", just messages).

Theory is that references will allow people trust each other and the network more and thus use it more, especially when they are new.

We also [considered](https://github.com/Trustroots/trustroots/issues/558) "personality tags" but decided against them.

**Comment by simison** [comment]
Here's a quick view on how leaving the reference could eventually look like:

![tr-reference-i1](https://user-images.githubusercontent.com/87168/47049952-650e9c00-d1a7-11e8-8e17-83d345bb507e.gif)

_(Sorry for odd tribe names, they're auto-generated with script...)_

**Comment by Regenwasser** [comment]
**Fantastic** - lot's of good stuff in all of this. And so much work! I heard and read often from people that they do not use the site, because of the lack of references so it's great to see this on _high priority_ on Github now as well. 

**My first question would be:** Where do things like priority order come from and how set are they? Can we maybe sort them with the help of users?

**The first thought overall:** Should we call iteration 1 a "reference"? 

To me a reference in this context, like a job reference or one left on user profiles on different or similiar sites (also those that most user are on), is a "_testimonial, character reference, recommendation, good word, backing_". But if it's _not_ including text then I assume could maybe lead to disappointment, because to most that maybe isn't what they expect when they think of reference or try to leave one. 

Yes, I am aware that this is _just_ the 1st iteration, but we'd still present it as "we launched references" on social media and the text write a reference is used here:
 
![image](https://user-images.githubusercontent.com/1645834/47052861-98065f00-d1aa-11e8-91ab-862d20d12131.png)

There is **no writing** involved in the 1st iteration. I also believe this should be considered, because we don't work on it full-time and other things come up it maybe is possible that the next iterations will be released many months later, so this might be our feature for a while.

The 1st iteration doesn't cover many use cases for references, but is maybe more of a form of social proof and an indication of "someone's experience/activity within the community". 

**So, with that in mind**: An alternative for the v1 text/feature name could be something _similiar to / in the spirit of_  "vouch for user" / "I met this user" or "report experience".
______________________________

**Also an issue to consider**: A negative "reference" (here the red "NO" displayed on the profile) without text can also have a very strong impact. So strong that I believe text is necessary.

Here some example to illustrate that a bit: 

> Julia leaves Peter a "No - would not meet/host again", because he brought a friend over at night without asking, disturbed the host in some way - or maybe just showed up just 20 min after the arranged time.
> 
> Now careful Tina sees Peters profile and won't meet him or won't bother answering his message, because the first person left him a "no, won't meet/host again". 

Text is necessary to explain the reason, because a "no" is very subjective. Always different to everyone. We can't explain when to say "No" and make sure people use it the right way. Some might laugh about negative references or just don't see certain things problematic, but a "No" or multiple ones without text is damning to many, which is why CS forces text.

So, an alternatives for a very basic 1st iteration would be: Only text _(skip "YES / NO / DON'T KNOW" labels - that's inside the text)_. 

Yes, only vouching is another possibility, but that would probably not add much to a profile as we already have links (FB, CS, etc.) and TR Contacts.

**Comment by simison** [comment]
> Where do things like priority order come from and how set are they? Can we maybe sort them with the help of users?

We've discussed big picture priorities with the board; I know there isn't a ton of transparency into it currently but we haven't really had volunteers neither. ;-)

**Comment by mrkvon** [comment]
Much complexity is here to prevent reciprocal negative references? (reference being unpublished first, limited time to leave a non-positive reference, inability to edit a reference)
How shall we deal with spam references, mistake references and references which change by repeated interaction between people?
This topic goes very complex very fast. Many possibilities.

**Comment by simison** [comment]
>Much complexity is here to prevent reciprocal negative references?

Exactly. Some of the patterns are pretty common now in different reference systems.

> How shall we deal with spam references, mistake references 

Case-by-case for starters to learn more and we can see next steps separately.

> references which change by repeated interaction between people?

This is an excellent question; would be nice to lay out different possibilities to compare cons&pros of each. 
One idea would be to allow adding new comments under the old the reference but keep the positive/negative flag. If the first impressions were negative, it's still valuable information even if their relationship later would turn positive.

**Comment by simison** [comment]
@Natttalia 

> First of all to give each user the posibility to turn on or off references on her/his own profile, empowering each person on the trustability that they want to generate (if we do this It would make sense to enable It by default

This is a really interesting idea I had not thought about or haven't seen elsewhere before.
How would you see it play out for people who receive (or would receive) negative references? Can you elaborate more on what kind of people/cases would likely turn references off for their profile?

A few questions to get deeper into feature's dynamics:

- Should we indicate in the profile (negatively) that they've turned off references, encouraging them to allow them?
- Should we allow hiding references after they've given or only let folks disallow giving them?
- Should members who have disabled their references be allowed to leave references to others?

**Comment by simison** [comment]
@Regenwasser 

>Also an issue to consider: A "reference" (here for example the "NO" somewhere on a profile) without text can also have a very strong impact. So strong that I believe text feedback is necessary.

This is an excellent point. I'll need to think about this a bit more but I'll get back to this. 👍 


>A reference in this context, one left on user profiles on different or similiar sites (also those that most user are on), is a "testimonial, character reference, recommendation, good word, backing" and consists usually of text. But if it's not possible to include text then I assume could maybe lead to disappointment, because to most that maybe isn't what they expect when they think of reference or try to leave one.

There's more to unwrap here but shortly — 

We wanted to try design references so that there would be no or little need for the free text but that they would still serve all the different purposes people use references for.
That's because on the other hand moderating references is really difficult and potentially lots of work that we cannot necessarily commit to (due nature of volunteering).

On the other hand we didn't feel like positive references as free text were never very useful or helpful. Value that 50 similar references give isn't a lot and adds a lot of noise to references. Think of _"all went well"_, _"they're great person"_ or listings of what they did.

That's why we tried to came up with structured ways to say useful things about the experience; like [personality tags](https://github.com/Trustroots/trustroots/issues/558) or vouching tribes. 

Structured reference data is nice because:
- we can summarize it at profile in a way that it's easier to digest and thus more useful
- we can sort search results based on it
- we can automate reactions to it ("people are leaving you negative references because X, you could change your behaviour X way"). Think of people who send extremely short messages to lots of people and getting getting voted down. We could suggest them to write longer messages.

**Comment by Natttalia** [comment]
Completely agree with Mika.
One of our priorities was to bring in all the benefits of the references
systems without the hassle of managing disputes. An even getting more
information from the references that what normally you can get from free
text.

Mainly disputes come from negative comments
It is true that on a negative comment we need to allow people to explain
themselves. Even on a positive comment there will be people wanting to
react. That is why I think we need to allow free text at the end.

So the idea could be to leave a reference following step by step system: so
first question, about tribes once filled, question about personality tags,
once filled, question about "would you recommend?", once this is filled a
free text question "any comments you want to add".
This system is how many social networks are collecting references, you go
step by step, so make sure you get answers in each section, so useful
information for us.
And finally if the receiver is not happy with a comment, can decide to
answer the comment or to not to show it. So no need for any disputes
managing, Receivers will decide in each case what they want to do.

About the allowing or not allowing references, it could work as in the
youtube comments system, you can decide at any time to disable it.

The question about if this person can leave references to others is
interesting. I would say that if you are not playing fair you should not
play, then you could be leaving negative references to everyone without any
consequences.
If we allow people to disable comments after a while it will be nice that
the references she/he left to others are hidden (maybe too complex
technically)


   - Should we indicate in the profile (negatively) that they've turned off
   references, encouraging them to allow them?
   - 👍

**Comment by simison** [comment]
>And finally if the receiver is not happy with a comment, can decide to answer the comment or to not to show it. So no need for any disputes managing, Receivers will decide in each case what they want to do.

Ah right, so the negative reference would be visible, just text explanation could be hidden by receiver?

**Comment by Regenwasser** [comment]
I think hiding negative reference won't work and if implemented correctly will not be used, because a hidden negative reference causes much greater concern to the person looking at a profile than a visible one.

**Comment by aoifehegarty** [comment]
I noticed that blablacar went from allowing users to leave an immediate reference, to requiring some time (48hrs I think) between the journey and the review of the driver/passenger.

I think whatever we do with references there should be a "cool off" period of a week after a stay before we allow references about a host or guest.

**Comment by Regenwasser** [comment]
> I´d also be quite happy to have individual tribes decide how they want to manage membership (ie invitation only or not) and references (negative allowed, must-show reference etc)

With references that doesn't work. Or every member would have to have seperate profiles for each tribe - or a member would only be allowed to be part of one Tribe.

**Comment by simison** [comment]
> I think whatever we do with references there should be a "cool off" period of a week after a stay before we allow references about a host or guest.

Interesting idea. Do you reckon the same results would come if we would:
- let the user leave in a reference (and it won't be public at this point yet)
- not make it public in first 48 hours even if the receiver responds to reference
- let both of them edit the reference until references go public
- send an email before 48 hours is up reminding that they can still change their reference if they want.

Probably not good for the first iteration but potentially for 2nd or 3rd.

**Comment by Regenwasser** [comment]
> There's some value in connecting references towards tribes though somehow, as we've done with "tribe vouches". Happy to elaborate more behind why tribes are important in the context of references. 

What's the problem that we want to solve with tribe vouches? 

I can't find anything in #208 in regards to that.

**Comment by aoifehegarty** [comment]
> > I think whatever we do with references there should be a "cool off" period of a week after a stay before we allow references about a host or guest.
> 
> Interesting idea. Do you reckon the same results would come if we would:
From a behavioural point of view, its better that no one writes references untill there has been a time to consider and integrate the experience. Often once we write something, we feel compelled to stick to our original position. 

I imagined that allowing editing or updates would be more complicated than simply having a workflow where the option of writing a ref is not presented for a certain amount of time. I quite like how BBC does this ie: 

Then you write a ref and you are not shown the other persons reference for you until either you submit a ref for the other person OR a timeout period is hit and you can no longer submit your ref (this is one month in BBC)

> Probably not good for the first iteration but potentially for 2nd or 3rd. Not saying we should 
> implement this as it might complicate the process (both for users and code-wise) for not _so_ much 
> value, but it's a good one to play around with.
I think having less impulsive reference writing is fairly important but agree its a nice to have rather than a critical fix.

**Comment by Regenwasser** [comment]
>I quite like how BBC does this ie: Then you write a ref and you are not shown the other persons reference for you until either you submit a ref for the other person OR a timeout period is hit and you can no longer submit your ref (this is one month in BBC)

Isn't what you describe here the same thing that Mikael also suggested? (See first post: "After writing the reference".)

The suggested "time until you write a reference"-method could only be tested and implemented if we had a request system with hosting dates.

**Comment by simison** [comment]
#### So we'll need email copy and short push notification copy for these cases:

- User A leaves user B a reference:
   - User B receives notification that they've received a reference.
    - User B cannot see it at this point but has 14 days time to leave a reference back. 
    - Should contain a link to leave the reference
    - See more elaborate details from https://github.com/Trustroots/trustroots/issues/718

- User B leaves a reference back to user A:
    - User A receives notification that they've received a reference from user B, mentioning that both references are now public.
    - Should contain a link to read the reference.
 
- User B did not leave the reference in return to user A yet, but there is still some time left. 
    - User B receivers reminder notification to leave the reference.
    - We'll probably send two of these; after half the time has passed (7/14 days) and when it's the last day.
    - Last email can sound urgent so that folks would feel like jumping in right-away to fill the reference.
    - Should contain a link to leave the reference

- User B did not leave the reference in return to user A, but since time window has passed, the original reference made by user A is now public.
    - Both users receive notification
    - Should contain a link to read the reference.

#### Notes:

- Links should not be inline within the text; instead they're a separate call-to-action button:
<img width="400" alt="image" src="https://user-images.githubusercontent.com/87168/47271758-bee6db80-d584-11e8-9f29-4b8173315187.png">

- See some of our [existing email templates](https://github.com/Trustroots/trustroots/tree/a6f5ae3864125a134f60c06f69003e8f501b0c91/modules/core/server/views/email-templates-text) as an example.

- Push notifications should be a sentence, max two.

**Comment by simison** [comment]
It's starting to look like we'll need freetext references sooner than I thought. It would be good to put some thinking to:
- How to make it seem like it's perfectly okay to leave references _without_ freetext. Having 50 positive references with _"They're great!"_ texts isn't really insightful or useful.
- Freetext considerations with negative references

**Comment by Regenwasser** [comment]
Maybe we should try to do some user research here before we continue.

> If it doesn't make to 1st iteration, then probably right after to the next one.

It seems at a minimum there needs to be text. So if the amount of feature is a roadblock maybe the @mrkvon comment makes sense here and the 1st iteration could just post references right away. The "complicated" double-blind references (with notification and everything) can be added after that. 

We don't have that many active users, so probably few retaliation references due to not too many negative experiences. Yes, the double-blind system is **very important** and also will lead to more people not being afraid of leaving something negative, but at first, we just want to get the feature out.

>  Having 50 positive references with "They're great!" texts isn't really insightful or useful.

That's a good point. It has a lot to do with how much we push the other person to sent a response. So, the copy (and amount) here is very important. It's about the balance of kindly reminding the user that no reference can be left after that deadline - but also make sure that nobody feels forced to write one. 

References are totally optional on TR and have to be started by one user, which means people don't leave them as often. We don't have hosting requests == Can't push both people == less useless references.  
 
But also, I'd guess it will probably take many months or years until just 1000 (so about 3%) of users reached so many references? So, that's enough time for iterations.

**Comment by mrkvon** [comment]
> The "complicated" double-blind references (with notification and everything) can be added after that.

**Comment by mrkvon** [comment]
Any ideas for the design for reading references in 1st iteration? Shall we show them as a list, or just a number and a type?

i.e. The number of references can be in a tab (like Contacts). Perhaps 3 numbers: green, gray, red.
When we click the tab, we may see a list of users who gave the references.
```
About    References (3 0 1)   Contacts (5)
---------------------------------------------------
user1    Guest, Met       2018-01-13    (in green)
user2    Host             2018-12-29    (in gray)
```

**Comment by nicksellen** [comment]
Yesterday we discussed this topic again ([minutes](https://wiki.trustroots.org/en/Meeting_Minutes/2020-01-13)), as completing https://github.com/Trustroots/trustroots/pull/918 is dependent on having a clearer idea of the references model we're using. Although the backend already has quite some implementation, it wasn't felt as the definitive way it should be.

@simison can maybe provide a bit more of the back story so far (as it's kind of long up there!), and @noahsmindfuck had offered to write up some of his thoughts we discussed in the call yesterday in more detail here.

If possible, it would be nice to be able to progress the topic a bit more in the next meeting (2020-01-27).

**Comment by noahsmindfuck** [comment]
I wonder how to construct the references in a way where they are not a value on a "moral" scale, to not have good/bad, positive/negative and certainly not a "neutral" option. 
Why?
For me the idea of a reference is to give a non-judgemental overview of an experience. Whether that is good/bad/positive/negative/whatever is always depending on the observer.
I mean, just because I didn't like the flat/hygiene/behaviour/political views/... of a person I shouldn't report them and I don't have to downvote them with a "negative" stamp but I can specify that I liked/didn't like without framing "that, what I like" as "good" and that, what I don't like" as "bad" by just putting it in words.

A binary option imho creates a moral code that is very intransparent and driven by the majority without leaving space for subculture, borderliners and minorities and I dislike that a lot :)

**Comment by mrkvon** [comment]
FYI The current implemetation screenshots:

### Create a reference

#### Step 1

![01_create_reference_know](https://user-images.githubusercontent.com/7449720/72669386-73460a80-3a31-11ea-9180-d8f2ff09554b.png)

#### Step 2

![02_create_reference_recommend_dontknow](https://user-images.githubusercontent.com/7449720/72669389-73dea100-3a31-11ea-9a86-b7407d57dba8.png)

#### Step 2 - report to moderators

![02_create_reference_recommend](https://user-images.githubusercontent.com/7449720/72669387-73dea100-3a31-11ea-9de7-2c26685d29f1.png)

#### Finished

![03_create_reference_submitted](https://user-images.githubusercontent.com/7449720/72669390-74773780-3a31-11ea-991d-7cfe7edef47b.png)

#### The other user receives an email notification to give a reference back

![04_reference_created_email_notification](https://user-images.githubusercontent.com/7449720/72669391-74773780-3a31-11ea-84d8-fceeb91c4500.png)

### List references

![05_list_references_self](https://user-images.githubusercontent.com/7449720/72669392-750fce00-3a31-11ea-886a-463ff5a336ca.png)

**Comment by mrkvon** [comment]
Thoughts about references (subjective):

## Trust and the purpose of references

References are here for 2 main purposes:
1. Learn about somebody I don't know: **Can I trust this person?**
2. Say thanks publicly

Trust may have multiple components (incomplete):
- how much time have we spent together?
- what's our relationship? (host/guest, close friend, family, acquaintance, romantic partner, collaborated together, ...)
- do I trust this person? (yes, not, a little, very much) - quantifiable...

Once I tried to implement this in [some application](https://ditup.org) (which was taking a lot of UI inspiration from trustroots).
![image](https://user-images.githubusercontent.com/7449720/72669641-c5d4f600-3a34-11ea-9e2b-8c88866a35ef.png)
This trust list would obviously need a remake to make sense in the context of trustroots.
There could be _I don't trust this person_, too. A flag for others about that user.

## Usefulness of written references

Written (free text) references can be difficult sometimes:
1. people don't really read them
2. the same stuff is repeated over and over again
3. judgmental

some thoughts about this:
1, 2: People could write free-text. Or they could click _I agree with this_ next to a reference of somebody else. This way the repetition could be skipped.
3: Let's ask people: "Please try to write subjectively, don't judge. Instead of writing _This person is funny_ write _We enjoyed a lot of laugh together_."

## Note

Does the above make sense? Does something need clarification?
I'm throwing these to the pot of ideas. Curious what comes out of the pot. :honey_pot: :dragon: :japanese_ogre:

**Comment by simison** [comment]
@mrkvon I really like your trail of thought and it aligns well what I've observed.

> 1. Learn about somebody I don't know: Can I trust this person?
> 2. Say thanks publicly

Totally. I'd also add "signal of experience" which doesn't necessarily signal _trust_ per-se but that they're familiar with guest/host concept and there are no surprises on that front, I know I'm for easy hosting when both sides are experienced.

Also one aspect that is often lost is that what is cool for someone can be horrible for someone else. Things like talkative/quiet, cleanliness/spottless, cheap/spends, etc. None of these are good or bad, just differences. Often hard to quantify (even with tags) so free text can help.

I'd say we likely want to de-signify the free text or tuck it hidden a bit exactly for reasons you said. It's still needed there it seems for possibility to explain nuance which otherwise can be lost in references.

**Comment by noahsmindfuck** [comment]
I really don't like the "I don't know" option :) I can already see many, many of them and you have no idea what that means.
Also, I think that the "yes" and "no" option should both have a "Why?" textfield so even if user A recommends host B bc they are friendly user C will not recomend bc they are annoying and you as a user D can then choose based on the yes/no and reasoning behind it whether you want to write them.

The free text should in my opinion be mandatory.

I would not combine the "no" reference with a report to moderation. Personally I don't want to deal with a lot of support messages bc. user F wouldn't recommend host G for having cats without saying it on the profile. That is information that the users need, not the moderators.
Although if a user H chooses the "no" option, after they sent the reference they should be asked whether they want to report that person to Trustroots staff for any reason.

**Comment by simison** [comment]
>  The free text should in my opinion be mandatory.

Could you expand on this? Curious to hear why you think so. :-)

I sense you refer to this (at https://github.com/Trustroots/trustroots/issues/718#issuecomment-574093206): _"A binary option imho creates a moral code that is very intransparent [...]"_?



> I would not combine the "no" reference with a report to moderation. Personally I don't want to deal with a lot of support messages bc. user F wouldn't recommend host G for having cats without saying it on the profile.

Could we help this by changing the wording? Like we've seen, users don't always report even theft directly to us. That's what I'd like to ensure with that feature and I share your worry that people would start sending us unrelevant things. If that happens tho, I think we can adjust (or even hide) the feature afterwards, but I'd like us to test it live for little while. What do you think?

> Although if a user H chooses the "no" option, after they sent the reference they should be asked whether they want to report that person to Trustroots staff for any reason.

I like this change, it does simplify the flow. 👍

**Comment by noahsmindfuck** [comment]
> The free text should in my opinion be mandatory.
Your reference is correct. Mandatory, because I think that is why we leave references. We want to create trust and needs transparency imho. If it is unclear how the yes/no vote came to pass. If I don't know the user that leaves the reference it actually doesn't tell me anything about the reference, just that a specific person that I don't know and don't trust per se either would/not recommend the host.

**Comment by nicksellen** [comment]
> If I don't know the user that leaves the reference it actually doesn't tell me anything about the reference, just that a specific person that I don't know and don't trust per se either would/not recommend the host.

This actually leads the way towards a network of trust, which seems how trust works in real life, so if you trust a person, and _they_ trust another person, that is most useful (trust * trust = trust), if a random person I don't trust then trusts someone it means much less (trust * nothing = ?).

I think having a graph of trust is probably a good direction to go for trustroots, and perhaps goes some way to address the scaling issue hospex sites seems to face (as they get large, they get more generic, less interesting).

How that might impact this feature decision now though, I'm less sure. It could be that by default you only see references made by your contacts (or their contacts), or by people you trusted (or one layer more), and if you select an option/tab, you get to see the references from _randoms_.

**Comment by simison** [comment]
>I think having a graph of trust is probably a good direction to go for trustroots,

+1. Quantifiable yes/no or positive/negative type answers have their negative aspects (that we're trying to soften quite a bit by providing more nuance in questions with different trust levels). At the same time, they're very powerful in building such a data network.

One of the early ideas was to give more weight on references from people who belong to the same tribes — either automatically in search or in some kind of visible score, and/or by giving folks a way to visualize it within the UI. In short, hitchhiker having a great experience with someone should matter more for other hitchhikers.

I don't think our network has necessarily enough variety for that to be useful, but some-day. :-) It's good to keep possibilities like that in mind.

**Comment by simison** [comment]
Something to consider about free-text:

One of the reasons we wanted to postpone implementing references was that it would likely cause more moderation work for volunteers. That is; people leaving negative references to each other, and complaining to us. We'd be in that kind of "subjectivity" state where it's already hard enough to do support or moderate language used in public references.

Not a blocker, but just something to consider and think if we can somehow direct/manage it by implementation details beforehand.

**Comment by nicksellen** [comment]
I had a half-baked idea to push this into the network-of-trust model.

A single question: _how much do you trust this person?_ with a range option (maybe 5 choices, from less to more, maybe with a negative option, or not).

Part of this idea is to address the concern in [this comment](https://github.com/Trustroots/trustroots/issues/718#issuecomment-576004670) from @noahsmindfuck that a rating from someone you don't know doesn't have much meaning. The logic for then having a textual reference is that it can provide more context, given you might not know them.

The trust ratings could be used in a way to show you a users trust score _in relation_ to people you trust yourself. As @Akronix mentions in our call this week, it's then a bit like googles pagerank algorithm, except TrustRank. @simison also mentioned that one of the _original_ conceptions of this was to do it _per tribe_, so to combine it with this rating idea, there would be a rating per tribe you have in common.

This would benefit from some mockups to make it more concrete.

Now, it could be this is just in addition to a text reference, and something that could be incrementally added.

It might be nice to seperate out what is the _spirit_ of this feature (what it is for, and where we want it to go), and the current implementation of it. We are kind of mixing those two discussions here, given there is already a developed and merged implementation.

I see a few practical options:
1. continue to discuss the _spirit_ and _goals_ of the references, until we can find some good harmony, then think about the rest later
2. someone propose some _small_ adjustments to what is already there, that mitigates some of the issues, but doesn't solve everything (makes it _acceptable_)
3. use what is already there, and get it live, then iterate on 2 or 1 later

I'm actually happy with any of those options.

@Akronix also made the good point that the survey answers might give some indication from the users how to proceed.

**Comment by Akronix** [comment]
I'd like to add my personal opinion:
I think it's good to release something to the general userbase only when it's sort of a definitive approach, otherwise we will too much influenced for whatever we early decide or else we will have to trash all the data we collected.
I'd proceed by continuing the discussion, but by using the feedback obtained by the user survey. I happily offer my self to analyse the data and do a comprehensive report on what the users think about this topic.*

I didn't mention it in the yesterday's meeting @simison, but in the Iberian collective we discussed about this concern that some people might prefer to not have references at all. We said that it could be an opt-out feature (people who opted-out won't be able to receive or to post references).

In a second thought, I think we can still do a beta version where people (or at least, us, the volunteers team) can test how the references feature would work in a early alpha stage. I think it's useful even though if it's only a small bunch of people, we can also see how much people sign-up for the beta testing and how much people use it. Plus, I imagine it's going in the same direction of what I was saying in the previous paragraph about leaving it as a optional feature.

Actually, I can remember that we already had a beta / playground version of trustroots to test references, @mrkvon? but can't remember what happened with it.

Lastly, in case you don't know, [this PageRank algorithm](https://en.wikipedia.org/wiki/PageRank) of ranking nodes of a network depending of how much is their actual influence in the network was initially created to sort the result pages of search queries in web searchers, but it's nowadays used in every different field and considered a general algorithm to be applied in social network analysis. So I can see a custom version applied for _"trust ranking"_.

 *A bit offtopic butt, don't we have a better channel for discussions than github issues? At least mediawiki? github feels too much technical and code-oriented for my taste. I am really missing Discourse for this :sweat:
